### ISSUE TITLE
System.Text.Json changes the data datatype from double to int on serialization

### ISSUE THREAD

**Issue by minato7** [dotnet/runtime#35195]
This is an issue found while upgrading to System.Text.Json in Asp.NET Core 3.1. We have two webservices which interact based on Json contract. 

When webservice1 based on ASP.NET Core, serializes using System.Text.Json a data with datatype **double with value 312.0** is passed to Python webservice, we expect serialization to have double value, but this is not the case with System.Text.Json.

This double value is serialized as int value which is causing our webservice2 to treat it as **int value 312** and causing variation in floating point results.

This was working perfectly fine with Newtonsoft.Json, as it treats double value 312.0 on serialization to be 312.0

I have uploaded the repo in this link:
[https://github.com/minato7/IssueSystemTextJsonDouble](url)

My assumption is if a contract has double datatype, it should be treated as double upon serialization and if it converts to int datatype this is an unexpected behavior, **there should be an option provided to treat this double like FloatParseHandling in Newtonsoft.Json** to achieve this behavior in System.Text.Json.

**Comment by weichch** [comment]
Some discussion on [original SO post](https://github.com/dotnet/runtime/issues/35195) for cross reference.

**Comment by Clockwork-Muse** [comment]
.... the first answer on SO points out: 

> What you have seen is a known portability issue with JSON:

.... so from that point of view there's nothing "wrong" with `System.Text.Json`...
What the answerer is pointing out is that JSON itself doesn't specify "type" for numbers, that's up to the programming language/api.

Ultimately, regardless of whether we adopt this suggestion, you should make your python code resilient in the face of this sort of input, especially if you may have external clients using this API.  If your API needs to do math on these numbers, it should be making sure it's deserialized them into the proper types first, not relied on them being formatted in a certain way.

**Comment by layomia** [comment]
The ["G17" standard format](https://docs.microsoft.com/dotnet/standard/base-types/standard-numeric-format-strings#the-general-g-format-specifier) is used by serializer when serializing `double` types. This format omits the decimal point and/or trailing zeros, if they are not required for the data to roundtrip.

Writing the smallest output required to round trip allows (de)serialization to be faster, and is employed by the serializer when when writing instances of other types, e.g. [`DateTime{Offset}`](https://docs.microsoft.com/en-us/dotnet/standard/datetime/system-text-json-support#support-for-formatting).

**Comment by minato7** [comment]
> The "G17" standard format is used by serializer when serializing double types. This format omits the decimal point and/or trailing zeros, if they are not required for the data to roundtrip.

This information is quite useful to understand what is happening under the hood, I have observed some truncation of double value coming from Python application and I could relate with G17 standard format.

**Comment by minato7** [comment]
> Ultimately, regardless of whether we adopt this suggestion, you should make your python code resilient in the face of this sort of input, especially if you may have external clients using this API.

@Clockwork-Muse Yes, we realized with this migration process to improve our input parsing, although our python web application is an internal microservice, we decided to make it resilient to any type of input.

**Comment by niemyjski** [comment]
This needs to be reopened.

**Comment by eiriktsarpalis** [comment]
Having taken a closer look, I agreed with the sentiment of https://github.com/dotnet/runtime/issues/35195#issuecomment-616958159 and https://github.com/dotnet/runtime/issues/35195#issuecomment-616639500. This is behaviour is by design and reflects the fact that IEEE floating points don't distinguish between `1`, `1.0`, `1E0`, or `1.000`. 

As a workaround, you could try writing a custom converter piggybacking on `decimal`, where decimal points are representable:

```C#
using System.Text.Json;
using System.Text.Json.Serialization;

JsonSerializerOptions options = new()
{
    Converters =
    {
        new DoubleWithDecimalPointConverter()
    }
};

Console.WriteLine(JsonSerializer.Serialize<double>(1, options)); // 1.0

public sealed class DoubleWithDecimalPointConverter : JsonConverter<double>
{
    public override double Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return reader.GetDouble();
    }

    public override void Write(Utf8JsonWriter writer, double value, JsonSerializerOptions options)
    {
        if (double.IsInteger(value))
        {
            writer.WriteNumberValue((decimal)value * 1.0M);
        }
        else
        {
            writer.WriteNumberValue(value);
        }
    }
}
```

**Comment by niemyjski** [comment]
@captainsafia Can we please get this one fixed?
